src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,82 +34,244 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
+        milestonesToBePopulated.addAll(likelyUnmatchedMilestones);
+    }
+
+    /**
+     * Adds a limited number of milestones to {@code milestonesToBePopulated}
+     * These milestones are taken from {@code allMilestones} if they match subsets of {@code userInput} and are
+     * not already in {@code milestonesToBePopulated}.
+     *
+     * Assumption: {@code limit} is a positive integer
+     * @param milestonesToBePopulated
+     * @param limit
+     * @param userInput
+     * @param allMilestones
+     */
+    private void addPartiallyMatchingMilestones(List<PickerMilestone> milestonesToBePopulated, int limit,
+                                                String userInput, List<PickerMilestone> allMilestones) {
+        assert limit > 0;
+        List<PickerMilestone> partiallyMatchingMilestones =
+                getLimitedMatchesFromPreviousInput(userInput, milestonesToBePopulated, limit, allMilestones);
+        milestonesToBePopulated.addAll(partiallyMatchingMilestones);
+    }
+
+    /**
+     * Adds a limited number of milestones to {@code milestonesToBePopulated}.
+     * These milestones are taken from {@code allMilestones} if its isMatching() returns true
+     *
+     * Assumption: {@code limit} is a positive integer
+     * @param milestonesToBePopulated
+     * @param limit
+     * @param allMilestones
+     */
+    private void addMatchingMilestones(List<PickerMilestone> milestonesToBePopulated, int limit,
+                                       List<PickerMilestone> allMilestones) {
+        assert limit > 0;
+        List<PickerMilestone> matchingMilestones = getLimitedMatchingMilestones(limit, allMilestones);
+        milestonesToBePopulated.addAll(matchingMilestones);
+    }
+
+    /**
+     * Gets the first {@code limit} milestones from {@code allMilestones} which are not already in
+     * {@code currentMilestones}.
+     *
+     * Assumptions: {@code limit} is not negative and {@code allMilestones} should already be ordered
+     * in the desired priority
+     * @param currentMilestones
+     * @param limit
+     * @param allMilestones
+     */
+    private List<PickerMilestone> getLimitedFirstMilestones(List<PickerMilestone> currentMilestones,
+                                                            int limit, List<PickerMilestone> allMilestones) {
+        return allMilestones.stream()
+                .filter(milestone -> !currentMilestones.contains(milestone))
+                .limit(limit)
                 .collect(Collectors.toList());
     }
 
-    private void filterMilestones(String query) {
-        currentMilestonesList
-                .forEach(milestone -> {
-                    boolean matchQuery = Utility.containsIgnoreCase(milestone.getTitle(), query);
-                    if (!milestone.isMatching()) milestone.setMatching(!matchQuery);
-                });
+    /**
+     * Gets a limited number of milestones from {@code allMilestones} matching previous userInputs(s)
```

Comment:
If the current userInput is of length 1 then there is only one previous userInput :stuck_out_tongue: